pkg/auth/handler/login_auth_url.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz/policy"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	"github.com/skygeario/skygear-server/pkg/core/inject"
+	"github.com/skygeario/skygear-server/pkg/core/server"
+)
+
+func AttachLoginAuthURLHandler(
+	server *server.Server,
+	authDependency auth.DependencyMap,
+) *server.Server {
+	server.Handle("/sso/{provider}/login_auth_url", &LoginAuthURLHandlerFactory{
+		authDependency,
+	}).Methods("OPTIONS", "POST")
+	return server
+}
+
+type LoginAuthURLHandlerFactory struct {
+	Dependency auth.DependencyMap
+}
+
+func (f LoginAuthURLHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &LoginAuthURLHandler{}
+	inject.DefaultInject(h, f.Dependency, request)
+
+	vars := mux.Vars(request)
+	h.ProviderName = vars["provider"]
+	providers := map[string]sso.Provider{
+		"google":    h.GoogleProvider,
+		"facebook":  h.FacebookProvider,
+		"instagram": h.InstagramProvider,
+		"linkedin":  h.LinkedInProvider,
+	}
+	if provider, ok := providers[h.ProviderName]; ok {
+		h.Provider = provider
+	}
+
+	return handler.APIHandlerToHandler(h, h.TxContext)
+}
+
+func (f LoginAuthURLHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return authz.PolicyFunc(policy.DenyNoAccessKey)
+}
+
+// LoginAuthURLRequestPayload login handler request payload
+type LoginAuthURLRequestPayload struct {
+	Scope       []string               `json:"scope"`
+	Options     map[string]interface{} `json:"options"`
+	CallbackURL string                 `json:"callback_url"`
+	RawUXMode   string                 `json:"ux_mode"`
+	UXMode      sso.UXMode
+}
+
+// Validate request payload
+func (p LoginAuthURLRequestPayload) Validate() error {
+	if p.CallbackURL == "" {
+		return skyerr.NewInvalidArgument("Callback url is required", []string{"callback_url"})
+	}
+
+	if p.UXMode == sso.Undefined {
+		return skyerr.NewInvalidArgument("UX mode is required", []string{"ux_mode"})
+	}
+
+	return nil
+}
+
+// LoginAuthURLHandler returns the SSO auth url by provider.
+//
+// curl \
+//   -X POST \
+//   -H "Content-Type: application/json" \
+//   -H "X-Skygear-Api-Key: API_KEY" \
+//   -d @- \
+//   http://localhost:3000/sso/<provider>/login_auth_url \
+// <<EOF
+// {
+//     "scope": ["openid", "profile"],
+//     "options": {
+//       "prompt": "select_account"
+//     },
+//     callback_url: <url>,
+//     ux_mode: <ux_mode>
+// }
+// EOF
+//
+// {
+//     "result": "<auth_url>"
+// }
+type LoginAuthURLHandler struct {
+	TxContext         db.TxContext           `dependency:"TxContext"`
+	GoogleProvider    sso.Provider           `dependency:"GoogleSSOProvider,optional"`
+	FacebookProvider  sso.Provider           `dependency:"FacebookSSOProvider,optional"`
+	InstagramProvider sso.Provider           `dependency:"InstagramSSOProvider,optional"`
+	LinkedInProvider  sso.Provider           `dependency:"LinkedInSSOProvider,optional"`
```

Comment:
Currently we will inject all the providers to the handler, that makes both the `DependencyMap` and all SSO handlers need to know the list of providers. When we want to add 1 new provider, we will need to update many places. Maybe we can inject one provider only by checking the route in `DependencyMap.Provide`? Handlers don't need to know about what providers we have.